### ISSUE TITLE
`random-sample` with `#:replacement? #f` does the wrong thing

### ISSUE THREAD

**Issue by MarcKaufmann** [racket/racket#3104]
My interpretation of `random-sample` is what the docs say:

> Returns a list of n elements of seq, picked at random, listed in any order. If replacement? is non-false, elements are drawn with replacement, which allows for duplicates.

This suggest that `(random-sample '(1 2) 2 #:replacement? #f)` should return `'(1 2)` or `'(2 1)` with equal probability. Instead it always returns `'(1 2)`. The reason is that it relies -- as the comments in the code say -- on https://rosettacode.org/wiki/Knuth%27s_algorithm_S. That is the wrong algorithm though, as that algorithm returns not a randomly ordered *list*, with all orderings equally likely. Instead it returns a *set* - or should be interpreted as such. And in that case `'(1 2)` and `'(2 1)` are of course identical. 

The easiest fix - although I am not sure how to efficiently implement this - is to randomize the order of the returned list. A dumb way would be to generate `n` different random numbers, associate the first with the first returned item etc, and then order the returned items by size of their associated random number. There must be algorithms that use fewer random numbers.

**Comment by rfindler** [comment]
The `shuffle` function in `racket/list` uses Fisher-Yates (an algorithm that implements a random without replacement function correctly).

**Comment by MarcKaufmann** [comment]
Thanks Robby, I am pretty sure I used `shuffle` in the past, but I totally forgot about it.

Would it be the right thing to add a `shuffle` call at the end of `random-sample` to fix it? If so, then I'll create the PR for that. 

If I do, how do I check that it passes all the tests etc?

**Comment by soegaard** [comment]
The docs on random-sample says:
    Returns a list of n elements of seq, picked at random, listed in any
order.

I read it as "n elements are picked simultaneously" and that the elements
of `seq` can list the elements in any order.
So I think  the wording of the documentation should be changed.

The purpose of Knuth's  algorithm S is: "To select n records at random from
a set of N, where 0<n<=N.".
And that fits with the alternative interpretation.

If we want `random-sample` to support the new interpretation, then I
suggest adding a new keyword argument,
say, #:unordered? to toggle between the current behaviour and the new.

/Jens Axel

**Comment by rfindler** [comment]
Changing the docs sounds better than changing the code to me! (I don't know the intention of the authors, however. Maybe @bennn or @stamourv does.)

**Comment by soegaard** [comment]
After reading about Algorithm S, I noticed that "Algorithm P (Shuffling)"
describes Fischer-Yates
and later that exercise 15 asks:

    15. [30] (Ole-Johan Dahl) If X_k=k for 1<=k<=t at the start of
algorithm P, and
    we terminate the algorithm when j reaches the value t-n, the sequence
X_{t-n+1},...,X_t
    is a random permutation of a random combination of n elements.
    Show how to simulate the effects of this procedure using only O(n)
cells of memory.

Is "a random permutation of a random combination of n elements" what you
wanted?

/Jens Axel



Den ons. 15. apr. 2020 kl. 20.58 skrev Jens Axel Søgaard <
jensaxel@soegaard.net>:

> The docs on random-sample says:
>     Returns a list of n elements of seq, picked at random, listed in any
> order.
>
> I read it as "n elements are picked simultaneously" and that the elements
> of `seq` can list the elements in any order.
> So I think  the wording of the documentation should be changed.
>
> The purpose of Knuth's  algorithm S is: "To select n records at random
> from a set of N, where 0<n<=N.".
> And that fits with the alternative interpretation.
>
> If we want `random-sample` to support the new interpretation, then I
> suggest adding a new keyword argument,
> say, #:unordered? to toggle between the current behaviour and the new.
>
> /Jens Axel
>
>
>

-- 
-- 
Jens Axel Søgaard

**Comment by MarcKaufmann** [comment]
Ah, I read the 'listed in any order' as 'any order is equally likely', but I see how it means 'any order is possible'. But this interpretation is a bit weird, since the wording means something different with and without replacement.

> Is "a random permutation of a random combination of n elements" what you
wanted?

If I understand you correctly, then yes it is.

While I can see why updating the docs might be better, optimally I feel that the either the functionality should change, or the name of the function should change. Sampling has a specific meaning in statistics -- at least I'm pretty sure it's not just my perception. In R (and I guess Python), drawing a sample by default means that the order matters. Moreover, the current version itself agrees that order matters when `#:replacement?  #t` -- but then suddenly order doesn't matter when `#:replacement?`. 

Of course I don't understand all the trade-offs - maybe this should be added to Rhombus for clean-up.

**Comment by bennn** [comment]
My interpretation matches Jens's

For rewording, maybe: "listed in an unspecified order."

The `#:unordered?` keyword sounds good if we want that behavior. (I think I've always wanted a set out of `random-sample`)

**Comment by MarcKaufmann** [comment]
Having thought more about it, the problem is that the there are two properties, resampling and order, and one keyword flips both together, yet the keyword refers to only one of the things it changes (see example below). The current function thus requires a description for two cases: the current description for `#:replacement? #t`, and the description that Ben suggests, plus a 'Thus not all permutations are equally likely' warning or similar (see R and Python below, others will think like me).

Based on that, I'd suggest:

- adding two cases to the docs that clarify it as per the above suggestion;
- ideally adding a new keyword to allow permutations. Not sure if `#:unordered?` is best, as I am not sure how to interpret it when reading it. Maybe #:random-order? or #:shuffled?. Not sure. 

Example on flipping both resampling and order property: 

If we interpret the docs as @bennn suggests, i.e. "listed in an unspecified order", then this describes the case for `#:replacement? #f` correctly. But with `#:replacement? #t`, that wording suggests that `'(1 1 2 2)` and `'(2 1 2 1)` are equally good, and the program doesn't guarantee that both are equally likely - which is not what the current code does I think.

Examples from R and Python:

- R has `sample()` (in base) and `sample_n()` (in dplyr) (https://dplyr.tidyverse.org/reference/sample.html)
- Python has `random.sample` and `numpy.random.choice` (https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html)

**Comment by soegaard** [comment]
Den tor. 16. apr. 2020 kl. 11.30 skrev Marc <notifications@github.com>:

> Having thought more about it, the problem is that the there are two
> properties, resampling and order, and one keyword flips both together, yet
> the keyword refers to only one of the things it changes (see example
> below).
>
The current function thus requires a description for two cases: the current
> description for #:replacement? #t, and the description that Ben suggests,
> plus a 'Thus not all permutations are equally likely' warning or similar
> (see R and Python below, others will think like me).
>
> Based on that, I'd suggest:
>
>    - adding two cases to the docs that clarify it as per the above
>    suggestion;
>    - ideally adding a new keyword to allow permutations. Not sure if
>    #:unordered? is best, as I am not sure how to interpret it when
>    reading it. Maybe #:random-order? or #:shuffled?. Not sure.
>
> Example on flipping both resampling and order property:
>
> If we interpret the docs as @bennn <https://github.com/bennn> suggests,
> i.e. "listed in an unspecified order", then this describes the case for #:replacement?
> #f correctly. But with #:replacement? #t, that wording suggests that '(1
> 1 2 2) and '(2 1 2 1) are equally good, and the program doesn't guarantee
> that both are equally likely - which is not what the current code does I
> think.
>
> Examples from R and Python:
>
>    - R has sample() (in base) and sample_n() (in dplyr) (
>    https://dplyr.tidyverse.org/reference/sample.html)
>    - Python has random.sample and numpy.random.choice (
>    https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html
>    )
>
> I agree that the part of the confusion is that "sample" is used in two
different contexts.

The most common example I know is getting a random sample of, say, size
n=1000 persons from a population of N=5700000.
Here the sample is thought of as a set.

The other is repeated sampling n times each sample has a size of 1 from a
list or set of size N.
Repeated sampling can with be with or without replacement. Here order is
important.

In Mathematica they have two different functions:

    RandomSample    (one single sample)
    RandomChoice    (repeated sampling)

Optional arguments allow for variations such as given weights to the
elements.

https://reference.wolfram.com/language/ref/RandomSample.html
https://reference.wolfram.com/language/ref/RandomChoice.html

Click "Details" to see the actual specification.

So I suggest we add a new `random-choice` instead of adding new keyword
arguments to `random-sample`?
This way the two concepts are separated into two separate functions.

/Jens Axel

**Comment by MarcKaufmann** [comment]
@soegaard : I read the Mathematica functions. Its RandomSample function does return a random permutation of a list if it is called with only that list, and I would guess that the same is true when you only get a sample of length 3 from a list of 5 items. Thus it would appear that Mathematica also does ensure that all permuations are equally likely, even when it returns a list where each element can only appear once. Thus it definitely doesn't run Knuth's S. 

What RandomSample vs RandomChoice do is that they are without replacement or with replacement - but both of them randomize orders. I would definitely not mind having `random-sample` and `random-choice` to drop the keyword, but it doesn't resolve the permutations problem which was my hang up. 

The change that requires no code update would be to change only the docs and tell people to use `shuffle` from `racket/list` when they want random permutations after sampling.

A code change that almost surely doesn't lead to any code breaking is to return a random permutation for `#:replacement? #f`, since that way people like me are not surprised, and everyone else was expecting some arbitrary sample anyway, the order not mattering. I mean, some people might rely on the fact that the only way that `(random-sample '(1 2 3 4 5 6) 5)` can return the set that doesn't have `4` is by returning `'(1 2 3 6 5)`, but it seems unlikely that existing code relies on this rather strange feature of the algorithm. They'd have to know it uses Knuth' S.

What you are suggesting Jens would maybe be worthwhile when doing a more thorough overhaul. Python also has `random.choice (at least in SciPy or Numpy or so), so a look at what other languages do and if they have converged for domain-specific reasons on one choice over another seems like the way to go. It would be less backwards compatible though.

Anyway, thanks for taking the time to respond, I am happy with any of the above, and happy to create an initial PR if someone decides what they think should be updated.

**Comment by MarcKaufmann** [comment]
Checking to see if someone with the authority to do so has a preference over the suggested possibilities. To recap:

1. Change only docs, having a separate explanation and example for value of `#:replacement?`
2. Update the function to shuffle the return arguments when `#:replacement? #f`. Shouldn't break any existing code, while making it work as all other `random-sample` implementations in other languages mentioned in this thread. (My preference.)
3. Have separate functions - `random-sample` and `random-choice` - for the different functionalities. (This would still confuse my past self if all permutations aren't equally likely.)

**Comment by samth** [comment]
Is it possible to do 2 directly, without first creating a random sample and then shuffling it?

**Comment by bennn** [comment]
I prefer 1 --- add an example & a margin-note for `shuffle` about permutations.

Doing 2 directly is going to be hard for sequences with side effects `(open-input-string "racket")`

**Comment by maueroats** [comment]
Doing (2) directly is "easy", a minimal change would be to start Knuth's Algorithm S with a shuffle of the first n items. (The whole sequence has to get forced even though the algorithm is "online", that is the issue you were alluding to?)

Coming to this discussion later, I do have a comment: it would be clearer if instead of just expressing a preference the responses also included at least a little bit of why certain behavior is preferred.

I'll post a detailed analysis to try to get commitment a change.

**Comment by maueroats** [comment]
TL;DR: 

* Random samples without replacement not appearing in a random order seems like a mistake to me.
* Python has a really clear statement about its sampling behavior (other languages, below).
* I checked GitHub and shuffling would break nobody's code. 
* Using the `random-sample` function with a type signature is a problem, maybe there is more mess here to fix. 

Long writeup from a boring trip...

I understand people are reluctant to touch old code.

I can also see that this issue would be *very* frustrating once it was discovered, especially with the (imo) unclear statement in the docs. 

## Restatement of the Issue

As I understand it, the main issue is that `random-sample` with `#:replacement? #f` produces a list that is not suitable to use to select 1 club president and 3 other club members in one call.

The documentation attempts to say this by using the phrase "taken in any order" to mean the list should be considered as an unordered set.

I think the unexpected behavior is suprising and inconsistent; please consider shuffle the output of `random-sample` in the non-replacement case.

##  Side Problem: Typed Racket Signature for `random-sample`

You [cannot specify the keyword argument to `random-sample` in Typed Racket](https://github.com/racket/typed-racket/issues/955). That means that sampling without replacement cannot be used from Typed Racket.  One fix would be to separate the with and without replacement functionality, so the polymorphic function no longer required a keyword argument.


## Other languages

Mathematica: `RandomChoice` (replacement) and `RandomSample` (no replacement). The `RandomSample` function does produces the samples in a random order.

Python 3: `random.choices` (replacement) and `random.sample` (no replacement). I personally found [the Python 3 `random` library documentation](https://docs.python.org/3/library/random.html#functions-for-sequences) to be very clearly stated. The most relevant quote from `random.sample` is

> The resulting list is in selection order so that all sub-slices will also be valid random samples. This allows raffle winners (the sample) to be partitioned into grand prize and second place winners (the subslices).

In two sentences, the documentation explains the behavior of the function and the motivation for that choice.

R (statistics language): has a multi-purpose `sample` function that is more closely related to the Racket one: it can sample with or without replacement, but the default is without replacement. Even sampling without replacement it produces a randomized list (so it is not just "pure Knuth S"). 


## Github usages of random-sample

I searched Github for `random-sample` in files with the `rkt` extension. There were 105 hits. The majority of these hits were from the same few sources (a class and a benchmark), and all of those selected with replacement.

### Uses without replacement

The uses of `random-sample` without replacement that I was able to identify do not rely on any particular order of the results.

* @stamourv: [dungeons and repossessions](https://github.com/stamourv/dungeons-and-repossessions)
* @Gram-Stone:  [Random sites in a plane for Vornoi diagram (apparently)](https://github.com/Gram-Stone/voronoi/blob/3fff3f893e4c2740346318cfab1d8adceef0e435/main.rkt).
* @nuprl: [Finds means of random samples](https://github.com/nuprl/gradual-typing-performance/blob/b7d19987ed736ac7b2eda6b31fe456f784695b6c/tools/summarize/summary.rkt
).
* @berry-cs: [Sinbad library](https://github.com/berry-cs/sinbad-rkt):  [Sorts immediately after sampling](https://github.com/berry-cs/sinbad-rkt/blob/2a31725861ce9a8c87bd4a3a38d355f571bdebcf/sinbad/plugin.rkt).
  
### Going to break with any change

There was one repository with tests that will break if anything is changed.

* @katsuya94 [racket-numerics](https://github.com/katsuya94/racket-numerics/blob/4dd0b7040579f909f36daa82ff60b14daf67315a/number.rkt)
  Tests of racket numerics using specific prng seeds and outcomes.
  Looks like an initial investigation.
  
  

## Appendix: Fast Sampling

I personally think a naive approach is sufficient, and lack of complexity is a plus, given the (looks non-existent?) framework in Racket for similar operations. This means the implementation (and documentation!) should focus on being correct, clear, and (imo) unsurprising. (I actually don't think Knuth's S fits the "clear" as in "obviously correct" parameter very well, but I imagine that the more you know the more clear it is.)

If performance is a consideration, then see [Vitter's Method D (offprint PDF)](http://www.ittc.ku.edu/~jsv/Papers/Vit87.RandomSampling.pdf) found via  [SO reference to Vitter](https://stackoverflow.com/questions/311703/algorithm-for-sampling-without-replacement). Someone could implement this algorithm from the original paper, but it looks like some nitty-gritty knowledge about doing things quickly would help and it is complex enough that some real testing is surely necessary.



  
P.S. John: whole name is fine. I sent you an email to confirm.
